server/controllers/finance/discounts.js

```diff
@@ -0,0 +1,176 @@
+/**
+* @module finance/discounts
+*
+* @desc The discounts API implements CRUD on the `discount` table in the
+* database.
+*
+* @requires lib/db
+*/
+
+var db = require('../../lib/db');
+
+/**
+ * @desc Looks up a discount in the database by its id.  Throws a prmoise error
+ * if the record does not exist, otherwise returns the record.
+ *
+ * @param {number} id
+ * @param {object} codes
+ * @returns {Promise} record
+ */
+function lookupDiscount(id, codes) {
+  'use strict';
+
+  var sql =
+    'SELECT d.id, d.label, d.description, d.inventory_uuid, ' +
+      'd.account_id, d.value, a.account_number, i.text as inventoryLabel ' +
+    'FROM discount AS d JOIN inventory AS i ON d.inventory_uuid = i.uuid ' +
+    'JOIN account AS a ON d.account_id = a.id ' +
+    'WHERE d.id = ?;';
+
+  return db.exec(sql, [ id ])
+  .then(function (rows) {
+
+    // if no matches in the database, throw a 404
+    if (rows.length === 0) {
+      throw codes.ERR_NOT_FOUND;
+    }
+
+
+    // return a single record
+    return rows[0];
+  });
+}
+
+/**
+ * GET /discounts/:id
+ *
+ * @desc Queries the database to a discount matching the param :id.  Expected
+ * to respond with either HTTP status 404 or 200.
+ */
+exports.detail = function detail(req, res, next) {
+  'use strict';
+
+  lookupDiscount(req.params.id, req.codes)
+  .then(function (discount) {
+    res.status(200).json(discount);
+  })
+  .catch(next)
+  .done();
+};
+
+/**
+ * GET /discounts
+ *
+ * @desc Queries the database to list all discounts recorded in the database.
+ * Returns HTTP status code 200 with an array containing zero or more records.
+ */
+exports.list = function list(req, res, next) {
+  'use strict';
+
+  var sql =
+    'SELECT d.id, d.label, d.value FROM discount AS d;';
+
+  db.exec(sql)
+  .then(function (rows) {
+    res.status(200).json(rows);
+  })
+  .catch(next)
+  .done();
+};
+
+/**
+ * POST /discounts
+ *
+ * @desc Inserts a single discount record into the database.  If the record
+ * is properly inserted, it returns a 201 CREATED status and the generated id.
+ */
+exports.create = function create(req, res, next) {
+  'use strict';
+
+  // expects the proposed record to be namespaced by "discount"
+  var data = req.body.discount;
+
+  var sql =
+    'INSERT INTO discount ' +
```

Comment:
True.  I prefer this syntax in this case because:
 1. I do not have to write `delete data.id`.  I'm lazy like that.
 2. I do not need the extreme flexibility that ? allows.  A discount only makes sense with each property defined.  A discount without a value doesn't make sense.  A discount without an inventory UUID doesn't make sense.  For something like a patient that has many optional properties, the ? syntax is a huge gain, but here I feel the flexibility is a danger rather than an advantage.